src/App.tsx

```diff
@@ -1,26 +1,324 @@
+      });
+  }
+
+  function deleteRequest(id: number) {
+    deleteTodo(id)
+      .then(() => {
+        setTimeout(() => {
+          setTodos(currentTodos =>
+            currentTodos.filter(_todo => _todo.id !== id),
+          );
+        }, 500);
+      })
+      .catch(() => {
+        setError('Unable to delete a todo');
+
+        setTimeout(() => setError(''), 3000);
+      })
+      .finally(() => {
+        setTimeout(() => {
+          setIdsOfRecedingTodos(ids => ids.slice(1));
+          inputRef.current?.focus();
+        }, 500);
+      });
+  }
+
+  function onDelete(id: number) {
+    setIdsOfRecedingTodos(ids => [...ids, id]);
+    deleteRequest(id);
+  }
+
+  function onDeleteSelected() {
+    todos.forEach(todo => {
+      if (todo.completed) {
+        const { id } = todo;
+
+        setIdsOfRecedingTodos(ids => [...ids, id]);
+        deleteRequest(id);
+      }
+    });
+  }
+
+  function onSelect(id: number, todo: Todo) {
+    setIdsOfTodosOnSelect(ids => [...ids, id]);
+
+    patchTodo(id, todo)
+      .then(_todo => {
+        setTodos(currentTodos => {
+          return currentTodos.map(t => (_todo.id === t.id ? _todo : t));
+        });
+        if (editedInputById !== -1) {
+          setEditedInputById(-1);
+        }
+      })
+      .catch(() => {
+        setError('Unable to update a todo');
+        setTimeout(() => setError(''), 3000);
+      })
+      .finally(() => setIdsOfTodosOnSelect(ids => ids.slice(1)));
+  }
+
+  function toogleAll() {
+    const todosForToogle = todos.filter(todo =>
+      allCompleted ? todo.completed : !todo.completed,
+    );
+
+    setAllCompleted(!allCompleted);
+
+    todosForToogle.forEach(todo => {
+      const { id, completed } = todo;
+
+      onSelect(id, { ...todo, completed: !completed });
+    });
+  }
+
+  function setForm(id: number) {
+    const editedTodo = todos.find(todo => todo.id === id);
+
+    if (editedTodo) {
+      setEditedValue(editedTodo.title);
+    }
+
+    setEditedInputById(id);
+  }
+
+  function keyUp(event: React.KeyboardEvent) {
+    setKey(event.key);
+  }
+
+  function onSubmitChangedInput(event: React.FormEvent) {
+    event.preventDefault();
+    const editedTodo = todos.find(todo => todo.id === editedInputById) as Todo;
+    const trimmedValue = editedValue.trim();
+    const { title, id } = editedTodo;
+
+    if (title === trimmedValue) {
+      setEditedInputById(-1);
+
+      return;
+    }
+
+    if (trimmedValue.length === 0) {
+      setIdsOfRecedingTodos(ids => [...ids, id]);
+      deleteRequest(id);
+
+      setTimeout(() => {
+        setEditedInputById(-1);
+      }, 500);
+
+      return;
+    }
+
+    editedTodo.title = trimmedValue;
+
+    onSelect(id, editedTodo);
+  }
+
+  // #endregion
+  // #region filter functions
+
+  function allSelected() {
+    setFilter(Filter.all);
+  }
+
+  function activeSelected() {
+    setFilter(Filter.active);
+  }
+
+  function completedSelected() {
+    setFilter(Filter.completed);
+  }
```

Comment:
it's redundant to create three functions and pass them into Footer, just pass the setFilter function